lib/chef/client.rb

```diff
@@ -676,10 +685,45 @@ def register(client_name = node_name, config = Chef::Config)
     # But is it already there?
     def check_certstore_for_key(cert_name)
       require "win32-certstore"
-      win32certstore = ::Win32::Certstore.open("MY")
+      if Chef::Config[:auth_key_registry_type] == "user"
+        # checking to see if this user is migrating from a cert in LocalMachine store
+        win32certstore = ::Win32::Certstore.open("MY")
+        existing_certificate = win32certstore.search("#{cert_name}")
+        if existing_certificate.nil? || existing_certificate.empty?
+          win32certstore = ::Win32::Certstore.open("MY", store_location: CERT_SYSTEM_STORE_CURRENT_USER)
+        else
+          update_key_and_register(cert_name)
+          delete_old_key(cert_name, CERT_SYSTEM_STORE_CURRENT_USER)
+        end
+      else
+        # checking to see if we're moving a user from the CurrentUser store
+        win32certstore = ::Win32::Certstore.open("MY", store_location: CERT_SYSTEM_STORE_CURRENT_USER)
+        existing_certificate = win32certstore.search("#{cert_name}")
+        if existing_certificate.nil? || existing_certificate.empty?
+          win32certstore = ::Win32::Certstore.open("MY")
+        else
+          update_key_and_register(cert_name)
+          delete_old_key(cert_name, CERT_SYSTEM_STORE_LOCAL_MACHINE)
+        end
+      end
       win32certstore.search("#{cert_name}")
     end
 
+    def delete_old_key(node, store)
+      self.class.delete_old_key(node, store)
+    end
+
+    def self.delete_old_key(node, store)
+      require "win32-certstore"
+      if store == CERT_SYSTEM_STORE_LOCAL_MACHINE
+        win32certstore = ::Win32::Certstore.open("MY")
+      else
+        win32certstore = ::Win32::Certstore.open("MY", store_location: CERT_SYSTEM_STORE_CURRENT_USER)
+      end
+      thumbprint = win32certstore.get_thumbprint("node")
+      win32certstore.delete(thumbprint)
+    end
+
```

Comment:
From our chat in the meeting. We don't want to delete a user key when requested to only look at system keys and vise-versa. Those are likely different runlist/nodenames.